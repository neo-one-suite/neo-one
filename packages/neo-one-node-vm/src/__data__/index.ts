import * as assets from './assets';
import * as transactions from './transactions';
import * as testUtils from './utils';

export * from './createBlockchain';
export * from './keys';

export { assets, testUtils, transactions };
