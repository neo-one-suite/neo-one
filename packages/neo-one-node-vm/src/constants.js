/* @flow */
import BN from 'bn.js';
import {
  type ExecutionAction,
  type TriggerType,
  type VMListeners,
  type WriteBlockchain,
} from '@neo-one/node-core';
import {
  type Block,
  type OpCode,
  type ScriptContainer,
  type SysCallName,
  type UInt160,
  type UInt160Hex,
  type VMState,
  common,
} from '@neo-one/client-core';
import type { Monitor } from '@neo-one/monitor';

import type { StackItem } from './stackItem';

export const MAX_STACK_SIZE = 2 * 1024;
export const MAX_INVOCATION_STACK_SIZE = 1024;
export const MAX_ARRAY_SIZE = 1024;
export const MAX_ITEM_SIZE = 1024 * 1024;
export const MAX_SCRIPT_LENGTH = 1024 * 1024;
export const MAX_VOTES = 1024;
export const BLOCK_HEIGHT_YEAR = 2000000;
export const MAX_ASSET_NAME_LENGTH = 1024;
const ratio = 100000;
export const FEES = {
  ONE: new BN(1 * ratio),
  TEN: new BN(10 * ratio),
  TWENTY: new BN(20 * ratio),
  ONE_HUNDRED: new BN(100 * ratio),
  TWO_HUNDRED: new BN(200 * ratio),
  FOUR_HUNDRED: new BN(400 * ratio),
  FIVE_HUNDRED: new BN(500 * ratio),
  ONE_THOUSAND: new BN(1000 * ratio),
};
export const FREE_GAS = common.TEN_FIXED8;

export type ExecutionStack = Array<StackItem>;
export type ExecutionInit = {|
  scriptContainer: ScriptContainer,
  triggerType: TriggerType,
  action: ExecutionAction,
  listeners: VMListeners,
  skipWitnessVerify: boolean,
  persistingBlock?: Block,
|};
type CreatedContracts = { [hash: UInt160Hex]: UInt160 };
export type Options = {|
  depth: number,
  stack: ExecutionStack,
  stackAlt: ExecutionStack,
  createdContracts: CreatedContracts,
  scriptHash: ?UInt160,
  entryScriptHash: UInt160,
|};
export type ExecutionContext = {|
  state: VMState,
  errorMessage?: string,
  blockchain: WriteBlockchain,
  init: ExecutionInit,
  engine: {
    run: (input: {| monitor: Monitor, context: ExecutionContext |}) => Promise<
      ExecutionContext,
    >,
    executeScript: (input: {|
      monitor: Monitor,
      code: Buffer,
      pushOnly?: boolean,
      blockchain: WriteBlockchain,
      init: ExecutionInit,
      gasLeft: BN,
      options?: Options,
    |}) => Promise<ExecutionContext>,
  },
  code: Buffer,
  pushOnly: boolean,
  scriptHash: UInt160,
  callingScriptHash: ?UInt160,
  entryScriptHash: UInt160,
  pc: number,
  depth: number,
  stack: ExecutionStack,
  stackAlt: ExecutionStack,
  gasLeft: BN,
  createdContracts: CreatedContracts,
|};

export const getResultContext = (context: ExecutionContext) => ({
  stack: context.stack,
  stackAlt: context.stackAlt,
  gasLeft: context.gasLeft,
  createdContracts: context.createdContracts,
});

export type OpResult = {|
  context: ExecutionContext,
  results?: Array<StackItem>,
  resultsAlt?: Array<StackItem>,
|};
export type OpInvokeArgs = {|
  monitor: Monitor,
  context: ExecutionContext,
  args: Array<StackItem>,
  argsAlt: Array<StackItem>,
|};
export type OpInvoke = (input: OpInvokeArgs) => Promise<OpResult> | OpResult;
export type Op = {|
  name: OpCode,
  in: number,
  inAlt: number,
  out: number,
  outAlt: number,
  modify: number,
  modifyAlt: number,
  invocation: number,
  array: number,
  item: number,
  fee: BN,
  invoke: OpInvoke,
  context: ExecutionContext,
|};

export type SysCall = {|
  name: SysCallName,
  in: number,
  inAlt: number,
  out: number,
  outAlt: number,
  modify: number,
  modifyAlt: number,
  invocation: number,
  array: number,
  item: number,
  fee: BN,
  invoke: OpInvoke,
  context: ExecutionContext,
|};
