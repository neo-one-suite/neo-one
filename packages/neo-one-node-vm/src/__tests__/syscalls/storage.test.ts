// tslint:disable no-object-mutation
import { crypto } from '@neo-one/client-common';
import { StorageFlags, utils } from '@neo-one/node-core';
import { AsyncIterableX } from '@reactivex/ix-es2015-cjs/asynciterable/asynciterablex';
import { BN } from 'bn.js';
import { runSysCalls, TestCase } from '../../__data__';
import { FEES } from '../../constants';
import { BufferStackItem, NullStackItem, StorageContextStackItem } from '../../stackItem';

const SYSCALLS: readonly TestCase[] = [
  {
    name: 'System.Storage.GetContext',
    result: ({ transaction }) => [new StorageContextStackItem(crypto.toScriptHash(transaction.script))],

    gas: FEES[400],
  },

  {
    name: 'System.Storage.GetReadOnlyContext',
    result: ({ transaction }) => [new StorageContextStackItem(crypto.toScriptHash(transaction.script), true)],

    gas: FEES[400],
  },

  {
    name: 'System.Storage.Get',
    result: [new BufferStackItem(Buffer.alloc(10, 1))],
    args: [
      {
        type: 'calls',
        calls: [
          {
            name: 'System.Storage.GetContext',
            type: 'sys',
          },
        ],
      },

      Buffer.alloc(1, 1),
    ],

    mockBlockchain: ({ blockchain }) => {
      blockchain.contract.get = jest.fn(async () => Promise.resolve({ manifest: { hasStorage: true } }));

      blockchain.storageItem.tryGet = jest.fn(async () => Promise.resolve({ value: Buffer.alloc(10, 1) }));
    },
    gas: FEES[1_000_000],
  },

  {
    name: 'System.Storage.Get',
    result: [new NullStackItem()],
    args: [
      {
        type: 'calls',
        calls: [
          {
            name: 'System.Storage.GetContext',
            type: 'sys',
          },
        ],
      },

      Buffer.alloc(1, 1),
    ],

    mockBlockchain: ({ blockchain }) => {
      blockchain.contract.get = jest.fn(async () => Promise.resolve({ manifest: { hasStorage: true } }));

      blockchain.storageItem.tryGet = jest.fn(async () => Promise.resolve(undefined));
    },
    gas: FEES[1_000_000],
  },

  {
    name: 'System.Storage.Find',
    result: () => (result) => {
      expect(result).toMatchSnapshot();
    },
    args: [
      {
        type: 'calls',
        calls: [
          {
            name: 'System.Storage.GetContext',
            type: 'sys',
          },
        ],
      },

      Buffer.alloc(1, 1),
    ],

    mockBlockchain: ({ blockchain }) => {
      blockchain.contract.get = jest.fn(async () => Promise.resolve({ manifest: { hasStorage: true } }));

      blockchain.storageItem.getAll$ = jest.fn(() => AsyncIterableX.of());
    },
    gas: FEES[1_000_000],
  },

  {
    name: 'System.Storage.AsReadOnly',
    result: ({ transaction }) => (stack) => {
      expect(stack.length).toEqual(1);
      // It should equal the call's script hash.
      expect(stack[0].value).not.toEqual(crypto.toScriptHash(transaction.script));

      expect(stack[0].isReadOnly).toBeTruthy();
    },
    args: [
      {
        type: 'calls',
        calls: [
          {
            name: 'System.Storage.GetContext',
            type: 'sys',
          },
        ],
      },
    ],

    gas: FEES[400],
  },

  {
    name: 'System.Storage.Put',
    result: [],
    args: [
      {
        type: 'calls',
        calls: [
          {
            name: 'System.Storage.GetContext',
            type: 'sys',
          },
        ],
      },

      Buffer.alloc(0, 0),
      Buffer.alloc(0, 0),
    ],

    mockBlockchain: ({ blockchain }) => {
      blockchain.contract.get = jest.fn(async () => Promise.resolve({ manifest: { hasStorage: true } }));

      blockchain.storageItem.tryGet = jest.fn(async () => Promise.resolve());
      blockchain.storageItem.add = jest.fn(async () => Promise.resolve());
    },
    gas: utils.ZERO,
  },

  {
    name: 'System.Storage.Put',
    result: [],
    args: [
      {
        type: 'calls',
        calls: [
          {
            name: 'System.Storage.GetContext',
            type: 'sys',
          },
        ],
      },

      Buffer.alloc(1024, 0),
      Buffer.alloc(0, 0),
    ],

    mockBlockchain: ({ blockchain }) => {
      blockchain.contract.get = jest.fn(async () => Promise.resolve({ manifest: { hasStorage: true } }));

      blockchain.storageItem.tryGet = jest.fn(async () => Promise.resolve());
      blockchain.storageItem.add = jest.fn(async () => Promise.resolve());
    },
    gas: new BN(102400000),
  },

  {
    name: 'System.Storage.Put',
    result: [],
    args: [
      {
        type: 'calls',
        calls: [
          {
            name: 'System.Storage.GetContext',
            type: 'sys',
          },
        ],
      },

      Buffer.alloc(1025, 0),
      Buffer.alloc(0, 0),
    ],

    mockBlockchain: ({ blockchain }) => {
      blockchain.contract.get = jest.fn(async () => Promise.resolve({ manifest: { hasStorage: true } }));

      blockchain.storageItem.tryGet = jest.fn(async () => Promise.resolve());
      blockchain.storageItem.add = jest.fn(async () => Promise.resolve());
    },
    gas: new BN(102500000),
    error: 'Item too large',
  },

  {
    name: 'System.Storage.Put',
    result: [],
    args: [
      {
        type: 'calls',
        calls: [
          {
            name: 'System.Storage.GetContext',
            type: 'sys',
          },
        ],
      },

      Buffer.alloc(0, 0),
      Buffer.alloc(1024, 0),
    ],

    mockBlockchain: ({ blockchain }) => {
      blockchain.contract.get = jest.fn(async () => Promise.resolve({ manifest: { hasStorage: true } }));

      blockchain.storageItem.tryGet = jest.fn(async () => Promise.resolve());
      blockchain.storageItem.add = jest.fn(async () => Promise.resolve());
    },
    gas: new BN(102400000),
  },
  {
    name: 'System.Storage.Put',
    result: [],
    args: [
      {
        type: 'calls',
        calls: [
          {
            name: 'System.Storage.GetContext',
            type: 'sys',
          },
        ],
      },

      Buffer.alloc(0, 0),
      Buffer.alloc(1024, 0),
    ],

    mockBlockchain: ({ blockchain }) => {
      blockchain.contract.get = jest.fn(async () => Promise.resolve({ manifest: { hasStorage: true } }));

      blockchain.storageItem.tryGet = jest.fn(async () => Promise.resolve({ manifest: { hasStorage: true } }));
      blockchain.storageItem.update = jest.fn(async () => Promise.resolve());
      blockchain.storageItem.add = jest.fn(async () => Promise.resolve());
    },
    gas: new BN(102400000),
  },

  {
    name: 'System.Storage.Put',
    result: [],
    args: [
      {
        type: 'calls',
        calls: [
          {
            name: 'System.Storage.GetContext',
            type: 'sys',
          },
        ],
      },

      Buffer.alloc(0, 0),
      Buffer.alloc(1025, 0),
    ],

    mockBlockchain: ({ blockchain }) => {
      blockchain.contract.get = jest.fn(async () => Promise.resolve({ manifest: { hasStorage: true } }));

      blockchain.storageItem.tryGet = jest.fn(async () => Promise.resolve());
      blockchain.storageItem.add = jest.fn(async () => Promise.resolve());
    },
    gas: new BN(102500000),
  },

  {
    name: 'System.Storage.Delete',
    result: [],
    args: [
      {
        type: 'calls',
        calls: [
          {
            name: 'System.Storage.GetContext',
            type: 'sys',
          },
        ],
      },

      Buffer.alloc(0, 0),
    ],

    mockBlockchain: ({ blockchain }) => {
      blockchain.contract.get = jest.fn(async () => Promise.resolve({ manifest: { hasStorage: true } }));
      blockchain.storageItem.tryGet = jest.fn(async () => Promise.resolve({ flags: StorageFlags.None }));
      blockchain.storageItem.delete = jest.fn(async () => Promise.resolve());
    },
    gas: FEES[1_000_000],
  },
];

describe('SysCalls: System.Storage', () => {
  runSysCalls(SYSCALLS);
});
