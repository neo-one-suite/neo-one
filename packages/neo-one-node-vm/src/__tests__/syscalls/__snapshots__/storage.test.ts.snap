// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SysCalls: System.Storage System.Storage.AsReadOnly 1`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.AsReadOnly 2`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.AsReadOnly 3`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.AsReadOnly 4`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Delete 1`] = `"contract.get"`;

exports[`SysCalls: System.Storage System.Storage.Delete 2`] = `
Array [
  Array [
    Object {
      "hash": "3f67626522f6e54e4e92ada1b1da8133409d5ec3",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Delete 3`] = `"storageItem.tryGet"`;

exports[`SysCalls: System.Storage System.Storage.Delete 4`] = `
Array [
  Array [
    Object {
      "hash": "3f67626522f6e54e4e92ada1b1da8133409d5ec3",
      "key": "",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Delete 5`] = `"storageItem.delete"`;

exports[`SysCalls: System.Storage System.Storage.Delete 6`] = `
Array [
  Array [
    Object {
      "hash": "3f67626522f6e54e4e92ada1b1da8133409d5ec3",
      "key": "",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Delete 7`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Delete 8`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Delete 9`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Delete 10`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Find 1`] = `
Array [
  IteratorStackItem {
    "mutableCount": 1,
    "referenceID": 1,
    "value": StackItemIterator {
      "enumerator": Object {
        "next": [Function],
        "return": [Function],
        "throw": [Function],
        Symbol(Symbol.asyncIterator): [Function],
      },
      "mutableDone": false,
    },
  },
]
`;

exports[`SysCalls: System.Storage System.Storage.Find 2`] = `"contract.get"`;

exports[`SysCalls: System.Storage System.Storage.Find 3`] = `
Array [
  Array [
    Object {
      "hash": "c5f1a32ca949050b5ad07587c475316c8b3c9647",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Find 4`] = `"storageItem.getAll$"`;

exports[`SysCalls: System.Storage System.Storage.Find 5`] = `
Array [
  Array [
    Object {
      "hash": "c5f1a32ca949050b5ad07587c475316c8b3c9647",
      "prefix": "01",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Find 6`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Find 7`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Find 8`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Find 9`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Get 1`] = `"contract.get"`;

exports[`SysCalls: System.Storage System.Storage.Get 2`] = `
Array [
  Array [
    Object {
      "hash": "c5f1a32ca949050b5ad07587c475316c8b3c9647",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Get 3`] = `"storageItem.tryGet"`;

exports[`SysCalls: System.Storage System.Storage.Get 4`] = `
Array [
  Array [
    Object {
      "hash": "c5f1a32ca949050b5ad07587c475316c8b3c9647",
      "key": "01",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Get 5`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Get 6`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Get 7`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Get 8`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Get 9`] = `"contract.get"`;

exports[`SysCalls: System.Storage System.Storage.Get 10`] = `
Array [
  Array [
    Object {
      "hash": "c5f1a32ca949050b5ad07587c475316c8b3c9647",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Get 11`] = `"storageItem.tryGet"`;

exports[`SysCalls: System.Storage System.Storage.Get 12`] = `
Array [
  Array [
    Object {
      "hash": "c5f1a32ca949050b5ad07587c475316c8b3c9647",
      "key": "01",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Get 13`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Get 14`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Get 15`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Get 16`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.GetContext 1`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.GetContext 2`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.GetContext 3`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.GetContext 4`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.GetReadOnlyContext 1`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.GetReadOnlyContext 2`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.GetReadOnlyContext 3`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.GetReadOnlyContext 4`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 1`] = `"contract.get"`;

exports[`SysCalls: System.Storage System.Storage.Put 2`] = `
Array [
  Array [
    Object {
      "hash": "09fa7552a1c818ce99853f0472cc114a0f230b2a",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Put 3`] = `"storageItem.tryGet"`;

exports[`SysCalls: System.Storage System.Storage.Put 4`] = `
Array [
  Array [
    Object {
      "hash": "09fa7552a1c818ce99853f0472cc114a0f230b2a",
      "key": "",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Put 5`] = `"storageItem.add"`;

exports[`SysCalls: System.Storage System.Storage.Put 6`] = `
Array [
  Array [
    StorageItem {
      "flags": 0,
      "hash": "09fa7552a1c818ce99853f0472cc114a0f230b2a",
      "key": "",
      "serializeWire": [Function],
      "sizeInternal": [Function],
      "value": "",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Put 7`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 8`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 9`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 10`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 11`] = `"contract.get"`;

exports[`SysCalls: System.Storage System.Storage.Put 12`] = `
Array [
  Array [
    Object {
      "hash": "62f76ec8eb103c633db2b67071fe277185a2dbe3",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Put 13`] = `"storageItem.tryGet"`;

exports[`SysCalls: System.Storage System.Storage.Put 14`] = `
Array [
  Array [
    Object {
      "hash": "62f76ec8eb103c633db2b67071fe277185a2dbe3",
      "key": "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Put 15`] = `"storageItem.add"`;

exports[`SysCalls: System.Storage System.Storage.Put 16`] = `
Array [
  Array [
    StorageItem {
      "flags": 0,
      "hash": "62f76ec8eb103c633db2b67071fe277185a2dbe3",
      "key": "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "serializeWire": [Function],
      "sizeInternal": [Function],
      "value": "",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Put 17`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 18`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 19`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 20`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 21`] = `"contract.get"`;

exports[`SysCalls: System.Storage System.Storage.Put 22`] = `
Array [
  Array [
    Object {
      "hash": "37cf61ca5211b65f1d13440c263db84c11a11c79",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Put 23`] = `"storageItem.tryGet"`;

exports[`SysCalls: System.Storage System.Storage.Put 24`] = `
Array [
  Array [
    Object {
      "hash": "37cf61ca5211b65f1d13440c263db84c11a11c79",
      "key": "",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Put 25`] = `"storageItem.add"`;

exports[`SysCalls: System.Storage System.Storage.Put 26`] = `
Array [
  Array [
    StorageItem {
      "flags": 0,
      "hash": "37cf61ca5211b65f1d13440c263db84c11a11c79",
      "key": "",
      "serializeWire": [Function],
      "sizeInternal": [Function],
      "value": "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Put 27`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 28`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 29`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 30`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 31`] = `"contract.get"`;

exports[`SysCalls: System.Storage System.Storage.Put 32`] = `
Array [
  Array [
    Object {
      "hash": "37cf61ca5211b65f1d13440c263db84c11a11c79",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Put 33`] = `"storageItem.tryGet"`;

exports[`SysCalls: System.Storage System.Storage.Put 34`] = `
Array [
  Array [
    Object {
      "hash": "37cf61ca5211b65f1d13440c263db84c11a11c79",
      "key": "",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Put 35`] = `"storageItem.update"`;

exports[`SysCalls: System.Storage System.Storage.Put 36`] = `
Array [
  Array [
    Object {
      "manifest": Object {
        "hasStorage": true,
      },
    },
    Object {
      "flags": 0,
      "value": "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Put 37`] = `"storageItem.add"`;

exports[`SysCalls: System.Storage System.Storage.Put 38`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 39`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 40`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 41`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 42`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 43`] = `"contract.get"`;

exports[`SysCalls: System.Storage System.Storage.Put 44`] = `
Array [
  Array [
    Object {
      "hash": "7971891d51c8bef0741ce6f53355dd3b6ae1aaf3",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Put 45`] = `"storageItem.tryGet"`;

exports[`SysCalls: System.Storage System.Storage.Put 46`] = `
Array [
  Array [
    Object {
      "hash": "7971891d51c8bef0741ce6f53355dd3b6ae1aaf3",
      "key": "",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Put 47`] = `"storageItem.add"`;

exports[`SysCalls: System.Storage System.Storage.Put 48`] = `
Array [
  Array [
    StorageItem {
      "flags": 0,
      "hash": "7971891d51c8bef0741ce6f53355dd3b6ae1aaf3",
      "key": "",
      "serializeWire": [Function],
      "sizeInternal": [Function],
      "value": "0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    },
  ],
]
`;

exports[`SysCalls: System.Storage System.Storage.Put 49`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 50`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 51`] = `Array []`;

exports[`SysCalls: System.Storage System.Storage.Put 52`] = `Array []`;
